src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,352 @@
+                    # Regex '[a]' (single char)
+                    builder.add_chars(hunichr(charset_value))
+                elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
+                    low, high = charset_value
+                    for char_code in hrange(low, high + 1):
+                        builder.add_chars(hunichr(char_code))
+                elif charset_code == sre.CATEGORY:
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
+                else:  # pragma: no cover
+                    # Currently there are no known code points other than
+                    # handled here. This code is just future proofing
+                    raise he.InvalidArgument(
+                        'Unknown charset code: %s' % charset_code
+                    )
+
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if context.flags & re.DOTALL:
+                return hs.characters()
+
+            return hs.characters(blacklist_characters='\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            if value == sre.AT_END:
+                return hs.one_of(hs.just(u''), hs.just(u'\n'))
+            return hs.just('')
+
+        elif code == sre.SUBPATTERN:
+            # Various groups: '(...)', '(:...)' or '(?P<name>...)'
+            old_flags = context.flags
+            if HAS_SUBPATTERN_FLAGS:  # pragma: no cover
+                # This feature is available only in specific Python versions
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = _strategy(value[-1], context)
+
+            context.flags = old_flags
+
+            if value[0]:
+                context.groups[value[0]] = strat
+                strat = hs.shared(strat, key=value[0])
+
+            return strat
+
+        elif code == sre.GROUPREF:
+            # Regex '\\1' or '(?P=name)' (group reference)
+            return hs.shared(context.groups[value], key=value)
+
+        elif code == sre.ASSERT:
+            # Regex '(?=...)' or '(?<=...)' (positive lookahead/lookbehind)
+            return _strategy(value[1], context)
+
+        elif code == sre.ASSERT_NOT:
+            # Regex '(?!...)' or '(?<!...)' (negative lookahead/lookbehind)
+            return hs.just('')
+
+        elif code == sre.BRANCH:
+            # Regex 'a|b|c' (branch)
+            return hs.one_of([_strategy(branch, context)
+                              for branch in value[1]])
+
+        elif code in [sre.MIN_REPEAT, sre.MAX_REPEAT]:
+            # Regexes 'a?', 'a*', 'a+' and their non-greedy variants
+            # (repeaters)
+            at_least, at_most, subregex = value
+            if at_most == 4294967295:
+                at_most = None
+            return hs.lists(_strategy(subregex, context),
```

Comment:
Ah ha - this is the common factor, all the failing tests feature a repeat somewhere.  I would guess the default `at_most` value for 'no limit' is different under 2.7.3; and `average_size` for the list is subsequently set to the mean of min and max size (see #217).   Try:

```
average = min([at_least + 5, (at_least + at_most) // 2])
if at_most > 10 ** 6:
    # Usually 2 ** 32, but that's way above the bound on our data size
    at_most = None
return hs.lists(_strategy(subregex, context),
                min_size=at_least, average_size=average, max_size=at_most
                ).map(''.join)
```